modules/markets-api/src/main/scala/org/ergoplatform/dex/markets/api/v1/services/AmmStats.scala

```diff
@@ -103,6 +114,59 @@ object AmmStats {
       } yield PoolSummary(poolId, pool.lockedX, pool.lockedY, tvl, volume, fees, yearlyFeesPercent)).value
     }
 
+    def getAvgPoolSlippage(poolId: PoolId, depth: Int): F[Option[PoolSlippage]] =
+      network.getCurrentHeight
+        .flatMap { currHeight =>
+          txr.trans(pools.trace(poolId, depth, currHeight))
+        }
+        .map { xs: List[PoolTrace] =>
+          if (xs.isEmpty)
+            None
+          else {
+            val slippageBySegment = xs
+              .groupBy(_.height)
+              .map { case (_, traces) =>
+                val max = traces.maxBy(_.gindex)
+                val min = traces.minBy(_.gindex)
+                val minPrice = RealPrice.calculate(
+                  min.lockedX.amount,
+                  min.lockedX.decimals,
+                  min.lockedY.amount,
+                  min.lockedY.decimals
+                )
+                val maxPrice = RealPrice.calculate(
+                  max.lockedX.amount,
+                  max.lockedX.decimals,
+                  max.lockedY.amount,
+                  max.lockedY.decimals
+                )
+                (maxPrice.value - minPrice.value).abs / (minPrice.value / 100)
+              }
+              .toList
+            Some(PoolSlippage(slippageBySegment.sum / slippageBySegment.size).scale(3))
+          }
+        }
```

Comment:
```suggestion
        .map {
          case Nil => None
          case xs =>
            val slippageBySegment = xs
              .groupBy(_.height)
              .map { case (_, traces) =>
                val max = traces.maxBy(_.gindex)
                val min = traces.minBy(_.gindex)
                val minPrice = RealPrice.calculate(
                  min.lockedX.amount,
                  min.lockedX.decimals,
                  min.lockedY.amount,
                  min.lockedY.decimals
                )
                val maxPrice = RealPrice.calculate(
                  max.lockedX.amount,
                  max.lockedX.decimals,
                  max.lockedY.amount,
                  max.lockedY.decimals
                )
                (maxPrice.value - minPrice.value).abs / (minPrice.value / 100)
              }
              .toList
            Some(PoolSlippage(slippageBySegment.sum / slippageBySegment.size).scale(3))
        }
```

Reply:
Btw, why not doing `.groupBy(_.height)` and `traces.maxBy(_.gindex)`/`traces.minBy(_.gindex)` in database?
Now you're loading a lot of data that is simply thrown away later.